src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,465 @@
+ *     a = a + 1
+ *     ~ y = ...
+ *     ~ z = ...
+ *     ~ w = ...
+ * }
+ *
+ * We get the vector F and matrix J
+ *
+ * F = [0,     J = [0, 4, 8,  12,
+ *      1,          1, 5, 9,  13,
+ *      2,          2, 6, 10, 14,
+ *      3]          3, 7, 11, 15]
+ *
+ * Where the numbers indicate their column-wise index. The solution replacement becomes:
+ *
+ * ~ x = ...  -> F[0] = ...
+ *               J[0] = ...
+ *               J[4] = ...
+ *               J[8] = ...
+ *               J[12] = ...
+ * a = a + 1
+ * ~ y = ...  -> ...
+ *
+ */
+class SympyReplaceSolutionsVisitor: public AstVisitor {
+  public:
+    enum class ReplacePolicy {
+        VALUE = 0,   //!< Replace statements matching by lhs varName
+        GREEDY = 1,  //!< Replace statements greedily
+    };
+    /// Empty ctor
+    SympyReplaceSolutionsVisitor() = delete;
+
+    /// Default constructor
+    SympyReplaceSolutionsVisitor(const std::vector<std::string>& pre_solve_statements,
+                                 const std::vector<std::string>& solutions,
+                                 const std::unordered_set<ast::Statement*>& to_be_removed,
+                                 const ReplacePolicy policy,
+                                 size_t n_next_equations);
+
+    /// idx (in the new statementVector) of the first statement that was added. -1 if nothing was
+    /// added
+    inline int replaced_statements_begin() const {
+        return replaced_statements_begin_;
+    }
+    /// idx (in the new statementVector) of the last statement that was added. -1 if nothing was
+    /// added
+    inline int replaced_statements_end() const {
+        return replaced_statements_end_;
+    }
+
+    void visit_statement_block(ast::StatementBlock& node) override;
+    void visit_diff_eq_expression(ast::DiffEqExpression& node) override;
+    void visit_lin_equation(ast::LinEquation& node) override;
+    void visit_non_lin_equation(ast::NonLinEquation& node) override;
+    void visit_binary_expression(ast::BinaryExpression& node) override;
+
+
+  private:
+    /** \brief Try to replace a statement
+     *
+     * @param node it can be Diff_Eq_Expression/LinEquation/NonLinEquation
+     * @param get_lhs method with witch we may get the lhs (in case we need it)
+     * @param get_rhs method with witch we may get the rhs (in case we need it)
+     */
+    void try_replace_tagged_statement(
+        const ast::Node& node,
+        const std::shared_ptr<ast::Expression>& get_lhs(const ast::Node& node),
+        const std::shared_ptr<ast::Expression>& get_rhs(const ast::Node& node));
+
+    /**
+     * \struct InterleavesCounter
+     * \brief Count interleaves of assignment statement inside the system of equations
+     */
+    struct InterleavesCounter {
+        /// Count an interleave if \ref in_system_ switches false -> true
+        void new_equation(const bool is_in_system);
+
+        /// Number of interleaves. We need to remove the first activation of the switch
```

Comment:
`the first activation of the switch` : may be just me that couldn't pick this up.

Reply:
improved the docs